### PR TITLE
Qualify diagnostic functions as pragma(printf)

### PR DESCRIPTION
**Pull request by nordlow**
Will improve type-checking of parameters used in C-`printf`-style diagnostics formatting.

### CODE REVIEW

**Comment by MoonlightSentinel** [comment]
This requires building with `-ignore` for older host compilers ( < 2.091.1 (?)).

**Comment by MoonlightSentinel** [comment]
Maybe do something like this:

```diff
diff --git a/src/build.d b/src/build.d
index 184b29313..62ec8f61a 100755
--- a/src/build.d
+++ b/src/build.d
diff --git a/src/build.d b/src/build.d
index 184b29313..76c6ec9f5 100755
--- a/src/build.d
+++ b/src/build.d
@@ -1007,6 +1007,11 @@ void processEnvironment()
     else
         env["HOST_DMD_KIND"] = "gdc";

+    // Determine version:
+    const verStart = hostDMDVersion.countUntil('v') + 1;
+    const versionNum = hostDMDVersion[verStart .. verStart + "X.XXX.X".length];
+    env["HOST_DMD_VERSION"] = versionNum;
+
     env["DMD_PATH"] = env["G"].buildPath("dmd").exeName;
     env.getDefault("DETAB", "detab");
     env.getDefault("TOLF", "tolf");
@@ -1023,6 +1028,10 @@ void processEnvironment()
     if (env["HOST_DMD_KIND"] != "gdc")
         dflags ~= ["-dip25"]; // gdmd doesn't support -dip25

+    // Ignore
+    if (versionNum < "2.092.0")
+        dflags ~= "-ignore";
+
     // TODO: add support for dObjc
     auto dObjc = false;
     version(OSX) version(X86_64)
```

**Comment by nordlow** [comment]
> This requires building with `-ignore` for older host compilers ( < ~2.091.1 (?)~ 2.092.0).

Should I add this flag? If so where?

**Comment by nordlow** [comment]
Does the error 

    src/dmd/errors.d(45): Error: pragma `printf` is missing a terminating `;`

mean that `-ignore` didn't have any effect in this case?

**Comment by MoonlightSentinel** [comment]
Found that `-ignore` did not work for unknown pragmas with bodys until 2.089.1 (fixed by [this PR](#10443)).

Guess we would have to implement another workaround for older host compilers or wait until they become recent enough...

**Comment by nordlow** [comment]
We could `version`-branch each definition of course.

**Comment by nordlow** [comment]
Is this the dmd version where `pragma(printf)` was introduced?

https://dlang.org/changelog/2.092.0.html

**Comment by MoonlightSentinel** [comment]
> Is this the dmd version where `pragma(printf)` was introduced?:
> https://dlang.org/changelog/2.092.0.html

Think so.

> **Update**: We could create wrapper functions that have different versions depending on `__VERSION__`...

Seems reasonable.

**Comment by nordlow** [comment]
What do you make of the error https://auto-tester.puremagic.com/show-run.ghtml?projectid=1&runid=4161908&isPull=true?

**Comment by MoonlightSentinel** [comment]
Running locally using `Ubuntu-WSL` (built with `-checkaction=context`):

```
core.exception.AssertError@src/dmd/lexer.d(2730): "escape hex sequence has 1 hex digits instead of 2" != "escape hex sequence has 1 hex digits instead of -703763272"       
----------------
??:? _d_assert_msg [0x7fa00f693bf2]
src/dmd/lexer.d:2730 nothrow @nogc bool dmd.lexer.__unittest_L2716_C1().expectDiagnosticHandler(ref const(dmd.globals.Loc), dmd.console.Color, const(char)*, const(char)*, core.internal.vararg.sysv_x64.__va_list_tag*, const(char)*, const(char)*) [0x7fa00f5b1d2f]
src/dmd/errors.d:338 nothrow void dmd.errors.verrorPrint(ref const(dmd.globals.Loc), dmd.console.Color, const(char)*, const(char)*, core.internal.vararg.sysv_x64.__va_list_tag*, const(char)*, const(char)*) [0x7fa00f5a1161]
src/dmd/errors.d:421 _Z6verrorRK3LocPKcP13__va_list_tagS3_S3_S3_ [0x7fa00f5a15c6]
src/dmd/errors.d:59 _Z6_errorRK3LocPKcz [0x7fa00f5a0309]
src/dmd/errors.d:53 _Z5errorRK3LocPKcz [0x7fa00f5a023e]
src/dmd/lexer.d:1230 nothrow dchar dmd.lexer.Lexer.escapeSequence(ref const(dmd.globals.Loc), ref const(char)*) [0x7fa00f5ae7a8]
src/dmd/lexer.d:2742 nothrow void dmd.lexer.__unittest_L2716_C1().test(immutable(char)[], immutable(char)[], dchar, uint) [0x7fa00f5b1da6]
src/dmd/lexer.d:2754 nothrow void dmd.lexer.__unittest_L2716_C1() [0x7fa00f5b17d6]
```

Win64 passes the test, so it looks like a parameter corruption.

File: src/dmd/errors.d [hunk 1]
```diff
@@ -74,7 +84,17 @@ extern (D) void error(Loc loc, const(char)* format, ...)
  *      format   = printf-style format specification
  *      ...      = printf-style variadic arguments
  */
-extern (C++) void error(const(char)* filename, uint linnum, uint charnum, const(char)* format, ...)
+static if (__VERSION__ < 2092)
+    extern (C++) void error(const(char)* filename, uint linnum, uint charnum, const(char)* format, ...)
+    {
+        _error(filename, linnum, charnum, format);
```
Comment (MoonlightSentinel) [inline src/dmd/errors.d:90]: This looks wrong ... isn't `va_list` required to forward c-style varargs?

Comment (MoonlightSentinel) [reply src/dmd/errors.d:90]: The test passes when using `va_list`:
```suggestion
        va_list ap;
        va_start(ap, format);
        verror(loc, format, ap);
        va_end(ap);
```

File: src/dmd/errors.d [hunk 2]
```diff
@@ -42,7 +42,17 @@ enum Classification
  *      format = printf-style format specification
  *      ...    = printf-style variadic arguments
  */
-extern (C++) void error(const ref Loc loc, const(char)* format, ...)
+static if (__VERSION__ < 2092)
+    extern (C++) void error(const ref Loc loc, const(char)* format, ...)
+    {
+        _error(loc, format);
```
Comment (MoonlightSentinel) [inline src/dmd/errors.d:48]: Using `va_list` here makes the test pass again.
```suggestion
        va_list ap;
        va_start(ap, format);
        verror(loc, format, ap);
        va_end(ap);
```

**Comment by nordlow** [comment]
So I guess I have to duplicate some code, then.

**Comment by nordlow** [comment]
Ready to go, @MoonlightSentinel?

File: src/dmd/errors.d (hunk 2 above)
Comment (nordlow) [reply src/dmd/errors.d]: Thanks a lot!

**Comment by nordlow** [comment]
> LGTM except that this overload is missing:
> 
> https://github.com/dlang/dmd/blob/ea93b5c79b2a1e4d198c6312bf5ba37fede9433a/src/dmd/errors.d#L62-L75
> 
> This overload cannot be marked with `pragma(printf)` as it's `extern(D)`so you would have to mark it as `extern(C++)`. But I think you can simply remove it because it's only used once:
> 
> https://github.com/dlang/dmd/blob/ace2aee91cb3e108172c74541c4b8527300ad3a6/src/dmd/dinifile.d#L346

The call in `dinifile.d` requires an r-value Loc. Should I store it as an r-value or add r-value overloads for

```D
extern (C++) void error(const Loc loc, const(char)* format, ...)
```

?

**Comment by MoonlightSentinel** [comment]
Rewrite it to store a local `Loc` for now, we can look into optimizations later.
 
```diff
-error(Loc(filename.xarraydup.ptr, lineNum, 0), "Use `NAME=value` syntax, not `%s`", pn); 
+Loc(filename.xarraydup.ptr, lineNum, 0)
+error(loc, "Use `NAME=value` syntax, not `%s`", pn); 
```

**Comment by nordlow** [comment]
> Rewrite it to store a local `Loc` for now, we can look into optimizations later.
> 
> ```diff
> -error(Loc(filename.xarraydup.ptr, lineNum, 0), "Use `NAME=value` syntax, not `%s`", pn); 
> +Loc(filename.xarraydup.ptr, lineNum, 0)
> +error(loc, "Use `NAME=value` syntax, not `%s`", pn); 
> ```

You mean do the transition from `const ref` to `in` at larger scale?

**Comment by MoonlightSentinel** [comment]
Yes, let's stick with the current default (`const ref`) and discuss optimizations in a dedicated PR.

**Comment by nordlow** [comment]
> Yes, let's stick with the current default (`const ref`) and discuss optimizations in a dedicated PR.

Done.

**Review by MoonlightSentinel — APPROVED** [review]

**Comment by nordlow** [comment]
Why does druntime fail to build in buildkite/dmd?

**Comment by MoonlightSentinel** [comment]
Seems like a spurious OOM, restarted the job manually.

```
Testing printf
--
  | ./generated/linux/debug/64/printf
  | /var/lib/buildkite-agent/builds/ci-agent-a5eb829d-1249-4ef6-a546-0eb660ed8d77-5/dlang/dmd/dmd/generated/linux/release/64/dmd -debug=LOGGING -m64 -fPIC -w -I../../src -I../../import -Isrc -defaultlib= -debuglib= -dip1000 -L-lpthread -L-lm -L/var/lib/buildkite-agent/builds/ci-agent-a5eb829d-1249-4ef6-a546-0eb660ed8d77-5/dlang/dmd/druntime/generated/linux/release/64/libdruntime.a -O -release -unittest -version=CoreUnittest -main -ofgenerated/linux/release/64/logging ../../src/gc/impl/conservative/gc.d ../../src/rt/lifetime.d
  | src/core/exception.d(647): [unittest] Memory allocation failed
  | core.exception.OutOfMemoryError@src/core/exception.d(647): Memory allocation failed
  | ----------------
  | 1/1 modules FAILED unittests
  | Makefile:15: recipe for target 'generated/linux/debug/64/printf.done' failed
  | make[2]: *** [generated/linux/debug/64/printf.done] Error 1


```

**Comment by nordlow** [comment]
> Seems like a spurious OOM, restarted the job manually.

Is dmd taking too much memory?

**Comment by MoonlightSentinel** [comment]
> Is dmd taking too much memory?

The OOM occurred in the `printf` executable (but I guess Buildkite just got too many jobs at once there)

**Comment by nordlow** [comment]
> The OOM occurred in the `printf` executable (but I guess Buildkite just got too many jobs at once there)

Is there no guarantee for the minimum amount of memory available?

**Comment by wilzbach** [comment]
> > The OOM occurred in the `printf` executable (but I guess Buildkite just got too many jobs at once there)
> 
> Is there no guarantee for the minimum amount of memory available?

Nope, we run the agents ourselves and to save resources we run multiple agents on one machine (and they share the memory). It usually works out well though ;-)

**Comment by nordlow** [comment]
What's up [this failure](https://auto-tester.puremagic.com/show-run.ghtml?projectid=1&runid=4162601&isPull=true)

**Comment by MoonlightSentinel** [comment]
> What's up with [this failure](https://auto-tester.puremagic.com/show-run.ghtml?projectid=1&runid=4162601&isPull=true)?

No idea but it looks environmental (`std.parallelism` wasn't changed in the last few days)

**Comment by nordlow** [comment]
> No idea but it looks environmental (`std.parallelism` wasn't changed in the last few days AFAICT)

Ok, thanks. Shall I do a forced push again?

**Comment by MoonlightSentinel** [comment]
No, I've restarted it manually

**Comment by nordlow** [comment]
> No, I've restarted it manually

Thanks.

**Comment by nordlow** [comment]
> No, I've restarted it manually

I presume I don't have the permissions to restart individual jobs manually, right?

**Comment by MoonlightSentinel** [comment]
> Can you restart parts of the autotester or only the whole?

Individual jobs (causing less workload than force-pushing)

> I presume I don't have the permissions to restart individual jobs manually, right?

No, the auto-tester requires a linked account with mege right IIRC. (But that situation should improve once we retire the auto-tester inf favour of other CI's, see e.g. #11653).

**Comment by wilzbach** [comment]
>No, the auto-tester requires a linked account with mege rights IIRC. (But that situation should improve once we retire the auto-tester inf favour of other CI's, see e.g. #11653).

Actually, anyone whose PRs can get tested on the auto-tester can also login there and deprecate the builds. Not that it would be important as the builds are constantly restarted. Once "auto-merge" is applied the PR branch will get higher priority and restart the build until it is green.

File: src/dmd/errors.d [hunk 3]
```diff
@@ -459,7 +585,17 @@ extern (C++) void vtip(const(char)* format, va_list ap)
...
+    extern (C++) void vdeprecationSupplemental(const ref Loc loc, const(char)* format, va_list ap)
+    {
+        _vdeprecationSupplemental(loc, format, ap);
+    }
+else
+    pragma(printf) extern (C++) void vdeprecationSupplemental(const ref Loc loc, const(char)* format, va_list ap)
+    {
+        _vdeprecationSupplemental(loc, format, ap);
+    }
+extern (C++) void _vdeprecationSupplemental(const ref Loc loc, const(char)* format, va_list ap)
```
Comment (wilzbach) [inline src/dmd/errors.d:598]: Should have been `private` and `extern(D)`

**Pull request merged**